### PR TITLE
[INLONG-7908][Sort] PostgreSQL connector supports parallel read

### LINKED ISSUE #7908: [Feature][Sort] PostgreSQL connector supports parallel read
### Description

_No response_

### InLong Component

InLong Sort

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by kuansix**
### Prepare a Pull Request 

- [INLONG-7908][Sort] Postgres connector supports parallel read
- Fixes #7908 

### Motivation

Enable incremental snapshot (support parallel read) and adapt metrics for PostgreSQL CDC.

#### 1. package `org.apache.inlong.sort.cdc.postgres.source`
- The core modification is implemented, refer to MySQL/Oracle's incremental snapshotting implementation.
- The feature can be enabled by setting `scan.incremental.snapshot.enabled=true`.

#### 2.package `io.debezium.connector.postgresql`
- Made some modifications to getting around the limitation.

#### 3.package `org.apache.inlong.sort.cdc.base`
- Add a new field List<String> schemaList and make it compatible with PostgreSQL and Debezium’s terminology.
- Add parameters related to incremental snapshot.
- Add parameters related to metrics.
- Adapt incremental snapshot and metrics.

### Documentation

*Documentation*

### CODE REVIEW

File: inlong-sort/sort-flink/cdc-base/src/main/java/org/apache/inlong/sort/cdc/base/source/IncrementalSource.java [hunk 1]
```diff
@@ -128,7 +128,7 @@ public IncrementalSourceReader<T, C> createReader(SourceReaderContext readerCont
                 .withRegisterMetric(RegisteredMetric.ALL)
                 .build();
 
-        sourceReaderMetrics.registerMetrics(metricOption, metricConfig.getMetricLabelList());
```
Comment (e-mhui) [inline inlong-sort/sort-flink/cdc-base/src/main/java/org/apache/inlong/sort/cdc/base/source/IncrementalSource.java:131]: It should not be deleted.

File: inlong-sort/sort-flink/cdc-base/src/main/java/org/apache/inlong/sort/cdc/base/debezium/table/RowDataDebeziumDeserializeSchema.java [hunk 2]
```diff
@@ -757,11 +757,6 @@ private void extractDdlRecord(SourceRecord record, Collector<RowData> out, Table
 
     }
 
-    @Override
```
Comment (e-mhui) [inline inlong-sort/sort-flink/cdc-base/src/main/java/org/apache/inlong/sort/cdc/base/debezium/table/RowDataDebeziumDeserializeSchema.java:760]: It should not be deleted.

**Comment by dockerzhang** [comment]
@kuansix, please add more description for this PR.

**Comment by EMsnap** [comment]
Please check the ut and compile thanks

**Comment by kuansix** [comment]
> @kuansix, please add more description for this PR.

done

**Comment by kuansix** [comment]
> Please check the ut and compile thanks

done

File: inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/postgres-cdc/src/main/java/io/debezium/connector/postgresql/PostgresObjectFactory.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+import java.nio.charset.StandardCharsets;
+import java.sql.SQLException;
+import java.time.Duration;
+import java.time.ZoneOffset;
+
+/**
+ * A factory for creating various Debezium objects
+ *
+ * <p>It is a hack to access package-private constructor in debezium.
```
Comment (EMsnap) [inline inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/postgres-cdc/src/main/java/io/debezium/connector/postgresql/PostgresObjectFactory.java:38]: Please add license for these files

File: inlong-sort/sort-flink/cdc-base/src/main/java/org/apache/inlong/sort/cdc/base/source/assigner/state/PendingSplitsStateSerializer.java [hunk 4]
```diff
@@ -369,8 +372,11 @@ private List<TableId> readTableIds(DataInputDeserializer in) throws IOException
         List<TableId> tableIds = new ArrayList<>();
         final int size = in.readInt();
         for (int i = 0; i < size; i++) {
+            // boolean useCatalogBeforeSchema = in.readBoolean();
```
Comment (EMsnap) [inline inlong-sort/sort-flink/cdc-base/src/main/java/org/apache/inlong/sort/cdc/base/source/assigner/state/PendingSplitsStateSerializer.java:375]: remove useless comment

File: inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/postgres-cdc/pom.xml [hunk 5]
```diff
@@ -29,14 +29,26 @@
...
-            <artifactId>flink-connector-postgres-cdc</artifactId>
+            <artifactId>flink-connector-debezium</artifactId>
+            <version>2.3.0</version>
         </dependency>
+        <dependency>
+            <groupId>com.ververica</groupId>
+            <artifactId>flink-cdc-base</artifactId>
+            <version>2.3.0</version>
+        </dependency>
+        <!--        <dependency>-->
```
Comment (EMsnap) [inline inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/postgres-cdc/pom.xml:47]: Useless comment here

File: inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/postgres-cdc/pom.xml [hunk 6]
```diff
@@ -29,14 +29,26 @@
...
+        <postgress.debezium.version>1.6.4.Final</postgress.debezium.version>
         <inlong.root.dir>${project.parent.parent.parent.parent.parent.basedir}</inlong.root.dir>
     </properties>
 
     <dependencies>
         <dependency>
             <groupId>com.ververica</groupId>
-            <artifactId>flink-connector-postgres-cdc</artifactId>
+            <artifactId>flink-connector-debezium</artifactId>
+            <version>2.3.0</version>
```
Comment (EMsnap) [inline inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/postgres-cdc/pom.xml:40]: please add version here

File: inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/postgres-cdc/pom.xml [hunk 7]
```diff
@@ -56,32 +68,59 @@
...
         <dependency>
             <groupId>io.debezium</groupId>
             <artifactId>debezium-connector-postgres</artifactId>
-            <version>${debezium.version}</version>
+            <version>${postgress.debezium.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.postgresql</groupId>
+            <artifactId>postgresql</artifactId>
+            <version>42.4.1</version>
```
Comment (EMsnap) [inline inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/postgres-cdc/pom.xml:84]: ditto

File: inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/postgres-cdc/src/main/java/org/apache/inlong/sort/cdc/postgres/source/options/SourceOptions.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+package org.apache.inlong.sort.cdc.postgres.source.options;
+
+import org.apache.inlong.sort.cdc.base.source.IncrementalSource;
+
+import org.apache.flink.configuration.ConfigOption;
+import org.apache.flink.configuration.ConfigOptions;
+
+/** Configurations for {@link IncrementalSource}. */
+public class SourceOptions {
```
Comment (EMsnap) [inline inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/postgres-cdc/src/main/java/org/apache/inlong/sort/cdc/postgres/source/options/SourceOptions.java:26]: source options is not used here

File: inlong-sort/sort-flink/cdc-base/src/main/java/org/apache/inlong/sort/cdc/base/config/JdbcSourceConfig.java [hunk 9]
```diff
@@ -49,6 +50,7 @@ public abstract class JdbcSourceConfig extends BaseSourceConfig {
     public JdbcSourceConfig(
             StartupOptions startupOptions,
             List<String> databaseList,
+            List<String> schemaList,
```
Comment (EMsnap) [inline inlong-sort/sort-flink/cdc-base/src/main/java/org/apache/inlong/sort/cdc/base/config/JdbcSourceConfig.java:53]: schemaList is never used

**Review by EMsnap — COMMENTED** [review]
Please add a test on parallel read and I guess some of the classes can reuse flink-cdc rather than copying them

File: inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/postgres-cdc/src/main/java/org/apache/inlong/sort/cdc/postgres/source/options/PostgresSourceOptions.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+/** Configurations for {@link PostgresSourceBuilder.PostgresIncrementalSource}. */
+public class PostgresSourceOptions extends JdbcSourceOptions {
+
+    public static final ConfigOption<Integer> PORT =
+            ConfigOptions.key("port")
+                    .intType()
+                    .defaultValue(5432)
+                    .withDescription("Integer port number of the PostgreSQL database server.");
+
+    public static final ConfigOption<String> DECODING_PLUGIN_NAME =
```
Comment (EMsnap) [inline inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/postgres-cdc/src/main/java/org/apache/inlong/sort/cdc/postgres/source/options/PostgresSourceOptions.java:38]: The config is never used

File: inlong-sort/sort-flink/cdc-base/src/main/java/org/apache/inlong/sort/cdc/base/relational/connection/JdbcConnectionPoolFactory.java [hunk 11]
```diff
@@ -47,6 +47,8 @@ public HikariDataSource createPooledDataSource(JdbcSourceConfig sourceConfig) {
         config.addDataSourceProperty(SERVER_TIMEZONE_KEY, sourceConfig.getServerTimeZone());
         config.setDriverClassName(sourceConfig.getDriverClassName());
 
+        // note: the following properties should be optional (only applied to MySQL)
+        config.addDataSourceProperty(SERVER_TIMEZONE_KEY, sourceConfig.getServerTimeZone());
```
Comment (gong) [inline inlong-sort/sort-flink/cdc-base/src/main/java/org/apache/inlong/sort/cdc/base/relational/connection/JdbcConnectionPoolFactory.java:51]: `config.addDataSourceProperty(SERVER_TIMEZONE_KEY, sourceConfig.getServerTimeZone());` is duplication code

File: inlong-sort/sort-flink/cdc-base/src/main/java/org/apache/inlong/sort/cdc/base/source/meta/offset/OffsetDeserializerSerializer.java [hunk 12]
```diff
@@ -84,7 +84,8 @@ default OffsetDeserializer createOffsetDeserializer() {
     default FinishedSnapshotSplitInfo deserialize(byte[] serialized) {
         try {
             final DataInputDeserializer in = new DataInputDeserializer(serialized);
-            TableId tableId = TableId.parse(in.readUTF());
+            boolean useCatalogBeforeSchema = in.readBoolean();
+            TableId tableId = TableId.parse(in.readUTF(), useCatalogBeforeSchema);
```
Comment (gong) [inline inlong-sort/sort-flink/cdc-base/src/main/java/org/apache/inlong/sort/cdc/base/source/meta/offset/OffsetDeserializerSerializer.java:88]: It should improve refer to https://github.com/ververica/flink-cdc-connectors/pull/2216 
```java
 default FinishedSnapshotSplitInfo deserialize(byte[] serialized) {
        try {
            final DataInputDeserializer in = new DataInputDeserializer(serialized);
            String tableIdStr = in.readUTF();
            String splitId = in.readUTF();
            Object[] splitStart = serializedStringToRow(in.readUTF());
            Object[] splitEnd = serializedStringToRow(in.readUTF());
            OffsetFactory offsetFactory = (OffsetFactory) serializedStringToObject(in.readUTF());
            Offset highWatermark = readOffsetPosition(in);
            boolean useCatalogBeforeSchema = true;
            if (in.available() > 0) {
                useCatalogBeforeSchema = in.readBoolean();
            }
            TableId tableId = TableId.parse(tableIdStr, useCatalogBeforeSchema);
            in.releaseArrays();

            return new FinishedSnapshotSplitInfo(
                    tableId, splitId, splitStart, splitEnd, highWatermark, offsetFactory);
        } catch (IOException e) {
            throw new FlinkRuntimeException(e);
        }
    }
```

File: inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/postgres-cdc/src/main/java/org/apache/inlong/sort/cdc/postgres/source/config/PostgresSourceConfig.java [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+        return new PostgresConnectorConfig(getDbzConfiguration());
+    }
+
+    public RelationalTableFilters getTableFilters() {
+        return getDbzConnectorConfig().getTableFilters();
+    }
+
+    @Override
+    public List<String> getMetricLabelList() {
+        return Arrays.asList(AbstractSourceInfo.DATABASE_NAME_KEY,
```
Comment (e-mhui) [inline inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/postgres-cdc/src/main/java/org/apache/inlong/sort/cdc/postgres/source/config/PostgresSourceConfig.java:103]: The key should be 'database', not 'db'.

File: inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/postgres-cdc/src/main/java/org/apache/inlong/sort/cdc/postgres/source/config/PostgresSourceConfig.java [hunk 14]
```diff
@@ -100,7 +101,7 @@ public RelationalTableFilters getTableFilters() {
 
     @Override
     public List<String> getMetricLabelList() {
-        return Arrays.asList(AbstractSourceInfo.DATABASE_NAME_KEY,
+        return Arrays.asList(DATABASE_NAME_KEY,
```
Comment (e-mhui) [inline inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/postgres-cdc/src/main/java/org/apache/inlong/sort/cdc/postgres/source/config/PostgresSourceConfig.java:104]: `cdc-base` module already has `DATABASE_NAME` constant, so you can import it, not define it by yourself.

image

you also refer Oracle CDC. 

image

**Comment by kuansix** [comment]
> Please add a test on parallel read and I guess some of the classes can reuse flink-cdc rather than copying them

1.add a test (done)

File: inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/oracle-cdc/src/main/java/org/apache/inlong/sort/cdc/oracle/source/config/OracleSourceConfigFactory.java [hunk 15]
```diff
@@ -108,10 +108,6 @@ public OracleSourceConfig create(int subtaskId) {
             props.setProperty("database.port", String.valueOf(port));
         }
 
-        if (schemaList != null) {
```
Comment (yunqingmoswu) [inline inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/oracle-cdc/src/main/java/org/apache/inlong/sort/cdc/oracle/source/config/OracleSourceConfigFactory.java:111]: Why change the code of Oracle cdc?

Comment (kuansix) [reply inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/oracle-cdc/src/main/java/org/apache/inlong/sort/cdc/oracle/source/config/OracleSourceConfigFactory.java:111]: schemaList is never used for pg  and oracle

File: inlong-sort/sort-flink/cdc-base/src/main/java/org/apache/inlong/sort/cdc/base/source/reader/IncrementalSourceReader.java [hunk 16]
```diff
@@ -237,7 +242,7 @@ private StreamSplit discoverTableSchemasForStreamSplit(StreamSplit split) {
     public void handleSourceEvents(SourceEvent sourceEvent) {
         if (sourceEvent instanceof FinishedSnapshotSplitsAckEvent) {
             FinishedSnapshotSplitsAckEvent ackEvent = (FinishedSnapshotSplitsAckEvent) sourceEvent;
-            LOG.debug(
+            LOG.info(
```
[the remark this answers is not in the document]
Comment (kuansix) [reply inlong-sort/sort-flink/cdc-base/src/main/java/org/apache/inlong/sort/cdc/base/source/reader/IncrementalSourceReader.java:245]: finished snapshot splits only once, and is necessary when debug online.

File: inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/postgres-cdc/src/main/java/org/apache/inlong/sort/cdc/postgres/source/reader/PostgresSourceRecordEmitter.java [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+
+import static com.ververica.cdc.connectors.base.source.meta.wartermark.WatermarkEvent.isHighWatermarkEvent;
+import static com.ververica.cdc.connectors.base.source.meta.wartermark.WatermarkEvent.isWatermarkEvent;
+import static com.ververica.cdc.connectors.base.utils.SourceRecordUtils.getHistoryRecord;
+import static com.ververica.cdc.connectors.base.utils.SourceRecordUtils.isDataChangeRecord;
+import static com.ververica.cdc.connectors.base.utils.SourceRecordUtils.isSchemaChangeEvent;
```
Comment (gong) [inline inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/postgres-cdc/src/main/java/org/apache/inlong/sort/cdc/postgres/source/reader/PostgresSourceRecordEmitter.java:51]: Fix dependency to `org.apache.inlong.sort.cdc`.

**Review by gong — APPROVED** [review]

**Review by dockerzhang — APPROVED** [review]

**Review by e-mhui — APPROVED** [review]

**Review by yunqingmoswu — APPROVED** [review]

**Pull request merged**